File: include/libqpageranges.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qpageranges.html
pub const qpageranges = struct {
    /// New constructs a new QPageRanges object.
    ///
    ///
    pub fn New() QtC.QPageRanges {
        return qtc.QPageRanges_new();
    }

    /// New2 constructs a new QPageRanges object.
    ///
    /// ``` other: QtC.QPageRanges ```
    pub fn New2(other: ?*anyopaque) QtC.QPageRanges {
        return qtc.QPageRanges_new2(@ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#operator-eq)
    ///
    /// ``` self: QtC.QPageRanges, other: QtC.QPageRanges ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPageRanges_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#swap)
    ///
    /// ``` self: QtC.QPageRanges, other: QtC.QPageRanges ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPageRanges_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#addPage)
    ///
    /// ``` self: QtC.QPageRanges, pageNumber: i32 ```
    pub fn AddPage(self: ?*anyopaque, pageNumber: i32) void {
        qtc.QPageRanges_AddPage(@ptrCast(self), @intCast(pageNumber));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#addRange)
    ///
    /// ``` self: QtC.QPageRanges, from: i32, to: i32 ```
    pub fn AddRange(self: ?*anyopaque, from: i32, to: i32) void {
        qtc.QPageRanges_AddRange(@ptrCast(self), @intCast(from), @intCast(to));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#toRangeList)
    ///
    /// ``` self: QtC.QPageRanges, allocator: std.mem.Allocator ```
    pub fn ToRangeList(self: ?*anyopaque, allocator: std.mem.Allocator) []QtC.QPageRanges__Range {
        const _arr: qtc.libqt_list = qtc.QPageRanges_ToRangeList(@ptrCast(self));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.QPageRanges__Range, _arr.len) catch @panic("qpageranges.ToRangeList: Memory allocation failed");
        const _data: [*]QtC.QPageRanges__Range = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#clear)
    ///
    /// ``` self: QtC.QPageRanges ```
    pub fn Clear(self: ?*anyopaque) void {
        qtc.QPageRanges_Clear(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#toString)
    ///
    /// ``` self: QtC.QPageRanges, allocator: std.mem.Allocator ```
    pub fn ToString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPageRanges_ToString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qpageranges.ToString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#fromString)
    ///
    /// ``` ranges: []const u8 ```
    pub fn FromString(ranges: []const u8) QtC.QPageRanges {
        const ranges_str = qtc.libqt_string{
            .len = ranges.len,
            .data = ranges.ptr,
        };
        return qtc.QPageRanges_FromString(ranges_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#contains)
    ///
    /// ``` self: QtC.QPageRanges, pageNumber: i32 ```
    pub fn Contains(self: ?*anyopaque, pageNumber: i32) bool {
        return qtc.QPageRanges_Contains(@ptrCast(self), @intCast(pageNumber));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#isEmpty)
    ///
    /// ``` self: QtC.QPageRanges ```
    pub fn IsEmpty(self: ?*anyopaque) bool {
        return qtc.QPageRanges_IsEmpty(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#firstPage)
    ///
    /// ``` self: QtC.QPageRanges ```
    pub fn FirstPage(self: ?*anyopaque) i32 {
        return qtc.QPageRanges_FirstPage(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#lastPage)
    ///
    /// ``` self: QtC.QPageRanges ```
    pub fn LastPage(self: ?*anyopaque) i32 {
        return qtc.QPageRanges_LastPage(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#detach)
    ///
    /// ``` self: QtC.QPageRanges ```
    pub fn Detach(self: ?*anyopaque) void {
        qtc.QPageRanges_Detach(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges.html#dtor.QPageRanges)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QPageRanges ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPageRanges_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qpageranges-range.html
pub const qpageranges__range = struct {
    /// New constructs a new QPageRanges::Range object.
    ///
    /// ``` other: QtC.QPageRanges__Range ```
    pub fn New(other: ?*anyopaque) QtC.QPageRanges__Range {
        return qtc.QPageRanges__Range_new(@ptrCast(other));
    }

    /// New2 constructs a new QPageRanges::Range object and invalidates the source QPageRanges::Range object.
    ///
    /// ``` other: QtC.QPageRanges__Range ```
    pub fn New2(other: ?*anyopaque) QtC.QPageRanges__Range {
        return qtc.QPageRanges__Range_new2(@ptrCast(other));
    }

    /// New3 constructs a new QPageRanges::Range object.
    ///
    ///
    pub fn New3() QtC.QPageRanges__Range {
        return qtc.QPageRanges__Range_new3();
    }

    /// New4 constructs a new QPageRanges::Range object.
    ///
    /// ``` param1: QtC.QPageRanges__Range ```
    pub fn New4(param1: ?*anyopaque) QtC.QPageRanges__Range {
        return qtc.QPageRanges__Range_new4(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    ///
    /// ``` self: QtC.QPageRanges__Range, other: QtC.QPageRanges__Range ```
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPageRanges__Range_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    ///
    /// ``` self: QtC.QPageRanges__Range, other: QtC.QPageRanges__Range ```
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPageRanges__Range_MoveAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpageranges-range.html#contains)
    ///
    /// ``` self: QtC.QPageRanges__Range, pageNumber: i32 ```
    pub fn Contains(self: ?*anyopaque, pageNumber: i32) bool {
        return qtc.QPageRanges__Range_Contains(@ptrCast(self), @intCast(pageNumber));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QPageRanges__Range ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPageRanges__Range_Delete(@ptrCast(self));
    }
};

File: src_original/libqpageranges.cpp
#include <QList>
#include <QPageRanges>
#define WORKAROUND_INNER_CLASS_DEFINITION_QPageRanges__Range
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qpageranges.h>
#include "libqpageranges.h"
#include "libqpageranges.hxx"

QPageRanges* QPageRanges_new() {
    return new QPageRanges();
}

QPageRanges* QPageRanges_new2(const QPageRanges* other) {
    return new QPageRanges(*other);
}

void QPageRanges_OperatorAssign(QPageRanges* self, const QPageRanges* other) {
    self->operator=(*other);
}

void QPageRanges_Swap(QPageRanges* self, QPageRanges* other) {
    self->swap(*other);
}

void QPageRanges_AddPage(QPageRanges* self, int pageNumber) {
    self->addPage(static_cast<int>(pageNumber));
}

void QPageRanges_AddRange(QPageRanges* self, int from, int to) {
    self->addRange(static_cast<int>(from), static_cast<int>(to));
}

libqt_list /* of QPageRanges__Range* */ QPageRanges_ToRangeList(const QPageRanges* self) {
    QList<QPageRanges::Range> _ret = self->toRangeList();
    // Convert QList<> from C++ memory to manually-managed C memory
    QPageRanges__Range** _arr = static_cast<QPageRanges__Range**>(malloc(sizeof(QPageRanges__Range*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QPageRanges::Range(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void QPageRanges_Clear(QPageRanges* self) {
    self->clear();
}

libqt_string QPageRanges_ToString(const QPageRanges* self) {
    QString _ret = self->toString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

QPageRanges* QPageRanges_FromString(const libqt_string ranges) {
    QString ranges_QString = QString::fromUtf8(ranges.data, ranges.len);
    return new QPageRanges(QPageRanges::fromString(ranges_QString));
}

bool QPageRanges_Contains(const QPageRanges* self, int pageNumber) {
    return self->contains(static_cast<int>(pageNumber));
}

bool QPageRanges_IsEmpty(const QPageRanges* self) {
    return self->isEmpty();
}

int QPageRanges_FirstPage(const QPageRanges* self) {
    return self->firstPage();
}

int QPageRanges_LastPage(const QPageRanges* self) {
    return self->lastPage();
}

void QPageRanges_Detach(QPageRanges* self) {
    self->detach();
}

void QPageRanges_Delete(QPageRanges* self) {
    delete self;
}

QPageRanges__Range* QPageRanges__Range_new(const QPageRanges__Range* other) {
    return new QPageRanges::Range(*other);
}

QPageRanges__Range* QPageRanges__Range_new2(QPageRanges__Range* other) {
    return new QPageRanges::Range(std::move(*other));
}

QPageRanges__Range* QPageRanges__Range_new3() {
    return new QPageRanges::Range();
}

QPageRanges__Range* QPageRanges__Range_new4(const QPageRanges__Range* param1) {
    return new QPageRanges::Range(*param1);
}

void QPageRanges__Range_CopyAssign(QPageRanges__Range* self, QPageRanges__Range* other) {
    *self = *other;
}

void QPageRanges__Range_MoveAssign(QPageRanges__Range* self, QPageRanges__Range* other) {
    *self = std::move(*other);
}

int QPageRanges__Range_From(const QPageRanges__Range* self) {
    return self->from;
}

void QPageRanges__Range_SetFrom(QPageRanges__Range* self, int from) {
    self->from = static_cast<int>(from);
}

int QPageRanges__Range_To(const QPageRanges__Range* self) {
    return self->to;
}

void QPageRanges__Range_SetTo(QPageRanges__Range* self, int to) {
    self->to = static_cast<int>(to);
}

bool QPageRanges__Range_Contains(const QPageRanges__Range* self, int pageNumber) {
    return self->contains(static_cast<int>(pageNumber));
}

void QPageRanges__Range_Delete(QPageRanges__Range* self) {
    delete self;
}
